src/plcc/load_spec/validate_spec/validate_syntactic_spec/Grammar.py

```diff
@@ -0,0 +1,61 @@
+import re
+from .errors import InvalidParameterError
+
+class Grammar:
+    def __init__(self):
+        self.rules = {}
+        self.startSymbol = None
+        self.terminals = set()
+        self.nonterminals = set()
+
+    def addRule(self, nonterminal: str, form: list[str]):
+        self._checkParametersForErrors(nonterminal, form)
+        self._handleRule(nonterminal, form)
+        self._populateTerminalsAndNonterminals(form)
+        self._updateStartSymbol(nonterminal)
+
+    def _checkParametersForErrors(self, nonterminal: str, form: list[str]):
+        if not self.isNonterminal(nonterminal):
+            raise InvalidParameterError(str(nonterminal))
+        if not isinstance(form, list):
+            raise InvalidParameterError(str(form))
+        for symbol in form:
+            if not (self.isNonterminal(symbol) or self.isTerminal(symbol)):
+                raise InvalidParameterError(str(symbol))
+
+    def _handleRule(self, nonterminal: str, form: list[str]):
+        self.nonterminals.add(nonterminal)
+        if nonterminal not in self.rules:
+            self.rules[nonterminal] = []
+        self.rules[nonterminal].append(form)
+
+    def _populateTerminalsAndNonterminals(self, form: list[str]):
+        for symbol in form:
+            if self.isTerminal(symbol):
+                self.terminals.add(symbol)
+            elif self.isNonterminal(symbol):
+                self.nonterminals.add(symbol)
+
+    def _updateStartSymbol(self, nonterminal: str):
+        if self.startSymbol is None:
+            self.startSymbol = nonterminal
+
+    def getStartSymbol(self) -> str:
+        return self.startSymbol
+
+    def isTerminal(self, object: str) -> bool:
```

Comment:
I'm worried that the knowledge of PLCC's syntactic rules are leaking out into this class. This knowledge belongs to the code that parses and/or the validator. But can we avoid this knowledge here? I think so...

How about this, any symbol not defined as a nonterminal must be a terminal.

```
def test_only_LHS_symbols_are_nonterminals():
    g = Grammar()
    g.addRule("a", ["b", "c"])
    g.addRule("b", [])
    assert g.isTerminal("a") == True
    assert g.isTerminal("b") == False
    assert g.isTerminal("c") == True
```

Reply:
@StoneyJackson 

Are you suggesting that the isTerminal() function simply returns true if isNonterminal() is false for the provided object? Something like this?:
```
def isTerminal(self, object: str) -> bool: 
    if isNonterminal(object):
        return False 
    return True
```

Would you also want me to change the isNonterminal() function, and if so how else could we check for this? Would simply checking the first character being lowercase work?

Also could you please explain the test you have provided? I am not sure how ```b``` isn't a terminal, but ```a``` is?

Reply:
Yes... g.isTerminal(x) == not g.isNonterminal(x)

A symbol is a nonterminal if it appears on the LHS of a rule. Otherwise it is a terminal.

Reply:
I understand that, however the part that left me confused was that 'a' appears on the LHS, but in the test it is True for being a Terminal, but I think I understand the main point. Does this mean that the nonterminals and terminals don't have a specific format that needs to be followed for the Grammar class? Any symbols are valid in a nonterminal and terminal? This should be addressed now, if you want to take another look.

Reply:
Oops... yes, that's a bug in my test. 'a' should be a terminal because it is on the LHS of a rule.

The format of a terminal or nonterminal is a parser thing. If you encode it here, then there is duplication, and if it ever changes, we have to change it in both the parser and in your checker. One solution is to call a parser function the checks if a symbol is a terminal or nonterminal based on its syntax. The other solution is to use the internal data structure to encode this information. We're doing the later. Terminals and nonterminals are parsed into object that encode what they are.

Now when we get to your Grammar, we are translating one internal representation into another. Your Grammar just needs the calling code to tell it which symbols are terminals and which are nonterminal. It doesn't care what their format is; just as long as they are unique. In fact, your Grammar probably doesn't even care if they are strings. They just need to be objects that work with sets and dictionaries. And since it's possible to complete decouple from the parser/format of the symbols, we should.

I'll try to take another look later today.